### PR TITLE
KAFKA-2678; partition level lag metrics can be negative

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

**Comment by lindong28** [comment]
To give an example, suppose the logEndOffset of the given partition on the leader and followers are 10, 8, 10. If the consumer reads uncommitted message from the follower with logEndOffset 10, then in the fetchResponse we may have:

```
partitionData.highWatermark - newOffset = 8 - 10 < 0
```

**Comment by guozhangwang** [comment]
LGTM. Just to clarify, this issue was only affecting replica fetcher's lag metrics; for normal consumers, since they are always consuming from leaders up to HW the lag metric should never be negative.

File: core/src/main/scala/kafka/server/AbstractFetcherThread.scala [hunk 1]
```diff
@@ -132,7 +132,7 @@ abstract class AbstractFetcherThread(name: String,
                       case None => currentPartitionFetchState.offset
                     }
                     partitionMap.put(topicAndPartition, new PartitionFetchState(newOffset))
-                    fetcherLagStats.getFetcherLagStats(topic, partitionId).lag = partitionData.highWatermark - newOffset
+                    fetcherLagStats.getFetcherLagStats(topic, partitionId).lag = Math.max(0, partitionData.highWatermark - newOffset)
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/AbstractFetcherThread.scala]: nit: we are implicitly casting 0 to 0L here.

Comment (lindong28) [inline core/src/main/scala/kafka/server/AbstractFetcherThread.scala]: @guozhangwang Thanks for your review. Problem fixed.

I tried to understand the how HW and messageOffset is updated but likely have missed understood the code. Thanks much for your clarification.

**Pull request closed without merging**